File: expo/superapp-mobile/modules/alrueya-secure-transport/ios/SecureTransportConfiguration.swift
import Foundation

struct SecureTransportConfiguration {
  let baseURL: URL?
  let environment: String
  let spkiPins: Set<String>

  var isProduction: Bool { environment == "production" }
  var isConfigured: Bool { isAllowedEndpoint && (!isProduction || !spkiPins.isEmpty) }
  var statusReason: String? {
    if !isAllowedEndpoint { return "endpoint_invalid" }
    if isProduction && spkiPins.isEmpty { return "production_pin_missing" }
    return nil
  }

  init(bundle: Bundle = .main) {
    guard
      let raw = bundle.object(forInfoDictionaryKey: "AlrueyaSecureTransportConfiguration") as? String,
      let data = raw.data(using: .utf8),
      let json = try? JSONSerialization.jsonObject(with: data),
      let object = json as? [String: Any],
      let urlString = object["baseUrl"] as? String,
      let parsedURL = URL(string: urlString.trimmingCharacters(in: .whitespacesAndNewlines))
    else {
      baseURL = nil
      environment = "development"
      spkiPins = []
      return
    }
    baseURL = parsedURL
    environment = object["environment"] as? String == "production" ? "production" : "development"
    let pins = (object["spkiPins"] as? [Any] ?? []).compactMap { $0 as? String }
    spkiPins = Set(pins.filter { $0.range(of: "^[A-Za-z0-9_-]{43}$", options: .regularExpression) != nil })
  }

  private var isAllowedEndpoint: Bool {
    guard let baseURL, let scheme = baseURL.scheme?.lowercased(), let host = baseURL.host?.lowercased(), !host.isEmpty else {
      return false
    }
    guard baseURL.user == nil, baseURL.password == nil, baseURL.query == nil, baseURL.fragment == nil else { return false }
    guard baseURL.path.isEmpty || baseURL.path == "/" else { return false }
    if scheme == "https" { return true }
    return environment == "development" && scheme == "http" && ["10.0.2.2", "127.0.0.1", "localhost"].contains(host)
  }
}

File: expo/superapp-mobile/modules/alrueya-secure-transport/android/src/main/java/online/alarabiya/securetransport/SecureTransportConfiguration.kt
package online.alarabiya.securetransport

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI

/**
 * Build-time configuration injected by the Expo config plugin. It is not a
 * credential: the endpoint and public SPKI hashes are safe to compile into an
 * app. The important invariant is that JavaScript cannot replace either one at
 * runtime, which would otherwise defeat certificate pinning.
 */
internal data class SecureTransportConfiguration(
  val baseUrl: String,
  val environment: String,
  val spkiPins: Set<String>,
) {
  val isProduction: Boolean get() = environment == "production"
  val isConfigured: Boolean
    get() = isAllowedEndpoint(baseUrl, environment) && (!isProduction || spkiPins.isNotEmpty())

  val statusReason: String?
    get() = when {
      !isAllowedEndpoint(baseUrl, environment) -> "endpoint_invalid"
      isProduction && spkiPins.isEmpty() -> "production_pin_missing"
      else -> null
    }

  companion object {
    private const val RESOURCE_NAME = "alrueya_secure_transport_configuration"
    private val loopbackHosts = setOf("10.0.2.2", "127.0.0.1", "localhost")
    private val pinPattern = Regex("^[A-Za-z0-9_-]{43}$")

    fun from(context: Context): SecureTransportConfiguration {
      val id = context.resources.getIdentifier(RESOURCE_NAME, "string", context.packageName)
      val raw = if (id == 0) null else context.getString(id)
      return parse(raw)
    }

    fun parse(raw: String?): SecureTransportConfiguration {
      if (raw.isNullOrBlank()) return SecureTransportConfiguration("", "development", emptySet())
      return try {
        val source = JSONObject(raw)
        val environment = if (source.optString("environment") == "production") "production" else "development"
        val baseUrl = source.optString("baseUrl").trim()
        val pins = source.optJSONArray("spkiPins")
          ?.strings()
          ?.filter(pinPattern::matches)
          ?.toSet()
          ?: emptySet()
        SecureTransportConfiguration(baseUrl, environment, pins)
      } catch (_: Exception) {
        SecureTransportConfiguration("", "development", emptySet())
      }
    }

    fun isAllowedEndpoint(baseUrl: String, environment: String): Boolean {
      val uri = runCatching { URI(baseUrl) }.getOrNull() ?: return false
      val host = uri.host?.lowercase()?.takeIf { it.isNotBlank() } ?: return false
      if (uri.userInfo != null || uri.query != null || uri.fragment != null) return false
      if (uri.path != null && uri.path !in setOf("", "/")) return false
      return when (uri.scheme?.lowercase()) {
        "https" -> true
        "http" -> environment == "development" && host in loopbackHosts
        else -> false
      }
    }

    private fun JSONArray.strings(): List<String> = buildList {
      for (index in 0 until length()) optString(index).takeIf { it.isNotBlank() }?.let(::add)
    }
  }
}
